services/src/main/java/org/keycloak/broker/spiffe/SpiffeIdentityProvider.java

```diff
@@ -0,0 +1,190 @@
+package org.keycloak.broker.spiffe;
+
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.core.UriInfo;
+import org.jboss.logging.Logger;
+import org.keycloak.broker.provider.AuthenticationRequest;
+import org.keycloak.broker.provider.BrokeredIdentityContext;
+import org.keycloak.broker.provider.ClientAssertionContext;
+import org.keycloak.broker.provider.ClientAssertionIdentityProvider;
+import org.keycloak.broker.provider.IdentityProvider;
+import org.keycloak.broker.provider.IdentityProviderDataMarshaller;
+import org.keycloak.common.util.Time;
+import org.keycloak.crypto.KeyWrapper;
+import org.keycloak.crypto.SignatureProvider;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.jose.jws.JWSHeader;
+import org.keycloak.jose.jws.JWSInput;
+import org.keycloak.keys.PublicKeyStorageProvider;
+import org.keycloak.keys.PublicKeyStorageUtils;
+import org.keycloak.models.FederatedIdentityModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.representations.JsonWebToken;
+import org.keycloak.services.Urls;
+import org.keycloak.sessions.AuthenticationSessionModel;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Implementation for https://datatracker.ietf.org/doc/draft-schwenkschuster-oauth-spiffe-client-auth/
+ *
+ * Main differences for SPIFFE JWT SVIDs and regular client assertions:
+ * <ul>
+*  <li><code>jwt-spiffe</code> client assertion type</li>
+ * <li><code>iss</code> claim is optional, uses SPIFFE IDs, which includes trust domain instead</li>
+ * <li><code>jti</code> claim is optional, and SPIFFE vendors re-use/cache tokens</li>
+ * <li><code>sub</code> is a SPIFFE ID with the syntax <code>spiffe://trust-domain/workload-identity</code></li>
+ * <li>Keys are fetched from a SPIFFE bundle endpoint, where the JWKS has additional SPIFFE specific fields (<code>spiffe_sequence</code> and <code>spiffe_refresh_hint</code>, the JWK does not set the <code>alg></code></li>
+ * </ul>
+ */
+public class SpiffeIdentityProvider implements IdentityProvider<SpiffeIdentityProviderConfig>, ClientAssertionIdentityProvider {
```

Comment:
The idea you proposed around separating the different aspects of the trust relationship with an identity provider should make this provider a lot simpler given that we won't need to override a lot of the methods here, right?

It should avoid decoupling the different types of providers and avoid, for instance, having `spifee` as an OAuth2-based broker. Or is that the case?

Reply:
Yeah, so SPIFFE should only implement the methods for client assertions and nothing else. Basically move all methods from IdentityProvider to UserIdentityProvider or something like that. We also need to figure out something around configuration as there's a ton of config options only applicable to user federation, but that are part of the core IdP model stuff.

Reply:
I'll create a follow-up issue for this; for now made a note here https://github.com/keycloak/keycloak/issues/42230